vmm-reference-interrupts.md

```diff
@@ -0,0 +1,116 @@
+
+There are five delivery modes for interrupts:
+* Fixed: The interrupt is sent to a specific VCPU.
+* SMI: System Management Interrupt.
+* NMI: Non-Maskable Interrupt. 
+* INIT: It is used to initialize the VCPU.
+* ExtINT: External Interrupt. Causes the processor to respond to the interrupt as if the interrupt originated in an 
+externally connected 
+
+Structure of 32-bit register `LVT` (Local Vector Table):
+* 8-bit vector
+* 3-bit delivery mode
+
+31 | 17 | 16 | 14 | 13 | 12 | 11 | 10 | 7 | 0
+--- | --- | --- | --- | --- | --- | --- | --- | --- | ---
+Reserved | Mask | Trigger Mode | Remote | interrupt input pin polarity | Delivery Status | Reserved | Delivery Mode | Vector 
+
+Refer to the [LAPIC documentation ](https://www.intel.com/content/dam/support/us/en/documents/processors/pentium4/sb/25366821.pdf) in intel's manual for more details. Figure 8.8 on Vol. 3A 8-17 shows the local vector table structure.
+
+The function `set_klapic_delivery_mode` sets the delivery mode from available delivery modes: `Fixed`, `SMI`, `NMI`, `INIT`, `ExtINT`. It uses `set_apic_delivery_mode` to get a `LVT` register and sets the delivery mode in it. The function takes 32 bit `reg` and `mode`. `mode` is shifted 8 bits to the left to place it at 11:8 bits and then last 8 bits of `reg` are ORed with `mode` to take only the last 8 bits of `reg`. The `LVT` structure looks as follows:
+
+31 | 17 | 16 | 14 | 13 | 12 | 11 | 10 | 7 | 0
+--- | --- | --- | --- | --- | --- | --- | --- | --- | ---
+00000000000000 | 0 | 00 | 0 | 0 | 0 | 0 | DELIVERY MODE | VECTOR 
+
+There are two helper functions defined to read and write to the LAPIC registers. The `read_le_i32` function reads from the LAPIC register at the given offset and returns the value read. 
+```rust
+pub fn get_klapic_reg(klapic: &kvm_lapic_state, reg_offset: usize) -> Result<i32> {
+    let range = reg_offset..reg_offset + 4;
+    let reg = klapic.regs.get(range).ok_or(Error::InvalidRegisterOffset)?;
+    Ok(read_le_i32(reg))
+}
+```
+
+The `write_le_i32` function writes the given value to the LAPIC register at the given offset.
+```rust
+pub fn set_klapic_reg(klapic: &mut kvm_lapic_state, reg_offset: usize, value: i32) -> Result<()> {
+    // The value that we are setting is a u32, which needs 4 bytes of space.
+    // We're here creating a range that can fit the whole value.
+    let range = reg_offset..reg_offset + 4;
+    let reg = klapic
+        .regs
+        .get_mut(range)
+        .ok_or(Error::InvalidRegisterOffset)?;
+    write_le_i32(reg, value);
+    Ok(())
+}
+```
+The helper functions `get_klapic_reg` uses the `read_le_i32` function to read from the LAPIC register at the given offset and returns the value read. The values are stored as `i32` which requires 4 `i8` values. Hence, the range is defined as `reg_offset..reg_offset + 4`. 
+
+```rust
+pub fn get_klapic_reg(klapic: &kvm_lapic_state, reg_offset: usize) -> Result<i32> {
```

Comment:
Code snippets for `get_klapic_reg` and `set_klapic_reg` look repeated

Reply:
Yes. Should've provided `read_le_i32` and `write_le_i32` instead.

Reply:
fixed